image_buffer/image_buffer.go

```diff
@@ -14,41 +15,61 @@ type CachedData struct {
 }
 
 type ImageBuffer struct {
-	mu             sync.Mutex
-	ringBuffer     []CachedData
-	toSend         []CachedData
-	captureTill    time.Time
-	lastCached     CachedData
-	windowSeconds  int
-	imageFrequency float64
-	maxImages      int
+	mu                  sync.Mutex
+	ringBuffer          []CachedData
+	toSend              []CachedData
+	captureFrom         time.Time
+	captureTill         time.Time
+	lastCached          CachedData
+	windowSecondsBefore int
+	windowSecondsAfter  int
+	imageFrequency      float64
+	maxImages           int
 }
 
-func NewImageBuffer(windowSeconds int, imageFrequency float64) *ImageBuffer {
+func NewImageBuffer(windowSeconds int, imageFrequency float64, windowSecondsBefore int, windowSecondsAfter int) *ImageBuffer {
 	// Calculate the maximum number of images to keep in the ring buffer
 	// Keep images for 2 * windowSeconds (before and after trigger)
-	maxImages := int(2 * float64(windowSeconds) * imageFrequency)
+	var maxImages int
+	if windowSeconds > 0 {
+		maxImages = int(2 * float64(windowSeconds) * imageFrequency)
+		windowSecondsBefore = windowSeconds
+		windowSecondsAfter = windowSeconds
+	} else {
+		maxImages = int(float64(windowSecondsBefore+windowSecondsAfter) * imageFrequency)
+	}
 	return &ImageBuffer{
-		ringBuffer:     []CachedData{},
-		toSend:         []CachedData{},
-		windowSeconds:  windowSeconds,
-		imageFrequency: imageFrequency,
-		maxImages:      maxImages,
+		ringBuffer:          []CachedData{},
+		toSend:              []CachedData{},
+		windowSecondsBefore: windowSecondsBefore,
+		windowSecondsAfter:  windowSecondsAfter,
+		imageFrequency:      imageFrequency,
+		maxImages:           maxImages,
 	}
 }
 
-func (ib *ImageBuffer) windowDuration() time.Duration {
-	return time.Second * time.Duration(ib.windowSeconds)
-}
-
 func (ib *ImageBuffer) MarkShouldSend(now time.Time) {
+	logger := logging.NewDebugLogger("client")
+
 	ib.mu.Lock()
 	defer ib.mu.Unlock()
 
-	ib.captureTill = now.Add(ib.windowDuration())
-
+	// Add images from the ring buffer that are within the window
+	beforeTimeBoundary := time.Second * time.Duration(ib.windowSecondsBefore)
+	afterTimeBoundary := time.Second * time.Duration(ib.windowSecondsAfter)
+	logger.Debugf("beforeTimeBoundary", beforeTimeBoundary.String())
+	logger.Debugf("afterTimeBoundary", afterTimeBoundary.String())
+
+	newCaptureFrom := now.Add(-beforeTimeBoundary)
+	newCaptureTill := now.Add(afterTimeBoundary)
+	if ib.captureTill.Before(now) {
+		ib.captureFrom = newCaptureFrom
```

Comment:
Put in a comment explaining how, if we're already in the middle of capturing new images, keep the old captureFrom. The comments should serve as a "director's commentary" explaining why we do the non-obvious things.